<a href="https://colab.research.google.com/github/yywu123/CSC44800----Artificial-Intelligence/blob/main/Assignment4%3A%20Judging%20Flowers/Assignment%204%20.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# <center> Assignment 4 :  Judging Flowers  </center>

![title](iris_flowers.jpeg)

In [341]:
# Imports and pip installations (if needed)
#pip install tabulate

from sklearn.datasets import load_iris
import pandas as pd 
import numpy as np
from tabulate import tabulate


from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier


# Part 1: Load the dataset


In [342]:
# Load the dataset (load remotely, not locally)
iris = load_iris()
print(iris.keys())
print(iris.target_names)



df = pd.DataFrame(data=np.c_[iris['data'], iris['target']], columns = iris['feature_names'] + ['Species'])


# Output the first 15 rows of the data
df.head(15)


dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])
['setosa' 'versicolor' 'virginica']


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    Species  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  
5       0.0  
6       0.0  
7       0.0  
8       0.0  
9       0.0  
10      0.0  
11      0.0  
12      0.0  
13      0.0  
14      0.0

In [343]:
# Display a summary of the table information (number of datapoints, etc.)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   Species            150 non-null    float64
dtypes: float64(5)
memory usage: 6.0 KB


In [344]:
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)     Species  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

## About the dataset


Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

This dataset consists information of 50 samples from each of three species of Iris (Iris setosa, Iris virginica and Iris versicolor). The dataset contain 150 record of Iris flower and for each record you can find information about sepal length, sepal width , petal length , petal width and species type. 

The feature of this dataset are sepal length, sepal width , petal length and petal width. The label is the species of the Iris flower, in this case the data only contain 0 , 1 , 2 and this three value will map to different species. 


- 0 --> setosa 
- 1 --> versicolor 
- 2 --> virginica

![title](iris-dataset.png)

# Part 2: Split the dataset into train and test


In [345]:
# Take the dataset and split it into our features (X) and label (y)


x = df[['sepal length (cm)','sepal width (cm)','petal length (cm)','petal width (cm)']]
y = df[['Species']]


columns = df.columns
print("Columns Name: ")
columns.tolist()


Columns Name: 


['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)',
 'Species']

In [346]:
## x contain sepal length, sepal width, petal length and petal width
x.head(5)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [347]:
## y contain the speciew value
y.head(5)

Species
0      0.0
1      0.0
2      0.0
3      0.0
4      0.0

In [348]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.1, train_size = 0.9, random_state=100)


In [349]:
## 90 % of dataset become train data  
X_train

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
97                 6.2               2.9                4.3               1.3
62                 6.0               2.2                4.0               1.0
77                 6.7               3.0                5.0               1.7
122                7.7               2.8                6.7               2.0
112                6.8               3.0                5.5               2.1
..                 ...               ...                ...               ...
87                 6.3               2.3                4.4               1.3
103                6.3               2.9                5.6               1.8
67                 5.8               2.7                4.1               1.0
24                 4.8               3.4                1.9               0.2
8                  4.4               2.9                1.4               0.2

[135 rows x 4 columns]

In [350]:
## 10 % of data become test data
X_test

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
128                6.4               2.8                5.6               2.1
11                 4.8               3.4                1.6               0.2
118                7.7               2.6                6.9               2.3
15                 5.7               4.4                1.5               0.4
123                6.3               2.7                4.9               1.8
135                7.7               3.0                6.1               2.3
32                 5.2               4.1                1.5               0.1
1                  4.9               3.0                1.4               0.2
116                6.5               3.0                5.5               1.8
45                 4.8               3.0                1.4               0.3
40                 5.0               3.5                1.3               0.3
115                6.4               3.2                5.3               2.3
26                 5.0               3.4                1.6               0.4
28                 5.2               3.4                1.4               0.2
145                6.7               3.0                5.2               2.3

In [351]:
y_train

Species
97       1.0
62       1.0
77       1.0
122      2.0
112      2.0
..       ...
87       1.0
103      2.0
67       1.0
24       0.0
8        0.0

[135 rows x 1 columns]

In [352]:
y_test

Species
128      2.0
11       0.0
118      2.0
15       0.0
123      2.0
135      2.0
32       0.0
1        0.0
116      2.0
45       0.0
40       0.0
115      2.0
26       0.0
28       0.0
145      2.0

# Part 3: Logistic Regression


In [353]:
# i. Use sklearn to train a LogisticRegression model on the training set

Logistic_Reg = LogisticRegression()
Logistic_Reg.fit(X_train, y_train.values.ravel())


LogisticRegression()

In [354]:
# ii. For a sample datapoint, predict the probabilities for each possible class
print("Sample datapoint information: ")
## pick the first data point as sample datapoint 
sample_datapoint = x[:1] 

##sample_datapoint with species 
sample_datapoint1 = df[:1]
sample_datapoint1

Sample datapoint information: 


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   

   Species  
0      0.0

In [355]:
## prediction
prediction = Logistic_Reg.predict_proba(sample_datapoint)[0]
data = {'Species': [0,1,2],
        'Species Name': ['Setosa', 'Versicolor', 'Virginica'],
        'Probability': ["{:.0%}".format(prediction[0]), "{:.0%}".format(prediction[1]), "{:.0%}".format(prediction[2])]}
print(tabulate(data, headers="keys", tablefmt='psql'))

print("Total Probability: " , prediction.sum())

+-----------+----------------+---------------+
|   Species | Species Name   | Probability   |
|-----------+----------------+---------------|
|         0 | Setosa         | 98%           |
|         1 | Versicolor     | 2%            |
|         2 | Virginica      | 0%            |
+-----------+----------------+---------------+
Total Probability:  1.0


Our prediction is correct. We can see that the correct species for the sample datapoint is 0, which is setosa. In our predxiction, setosa have 97% probability to ve the species for the sample datapoint 

In [356]:
# iii. Report on the score for Logistic regression model, what does the score measure?

train_score = Logistic_Reg.score(X_train,y_train)
test_score = Logistic_Reg.score(X_test, y_test)
print("The train score for Logistic Regression model: ",train_score)
print("The test score for Logistic Regression model ",test_score)

The train score for Logistic Regression model:  0.9777777777777777
The test score for Logistic Regression model  1.0


 As we know, the score() function from sklearn will return the mean accuracy on the given test data and labels. The best possible you can get is 1, which means the model can pefectly predict the label. Therefore, if the score is really close to 1, it means your models is very good and accurate at predicting the label. 

Our logistic regression model got the score of 0.9777 on the train date. The score is really close to 1, which means that our model is nearly perfect on predicting the species of the Iris during training. Based on the training data, the accuracy of this model is 97.7 % . 

Our logistic regression model got the score of 1 on the testing data. This means that our model perfectly predict the species of the Iris for all the testing dataset. This is impressive result that we get from our model. The score that we get indicate that our model is perfect on predicting the Iris species.

In [357]:
# iv. Extract the coefficents and intercepts for the boundary line(s)

coefficent = Logistic_Reg.coef_
intercept = Logistic_Reg.intercept_


print("Coefficent: ", coefficent)
print("Intercept: ", intercept)

Logistic_Reg_df = pd.DataFrame(data=np.c_[coefficent, intercept], columns = iris['feature_names'] + ['Intercept'])
Logistic_Reg_df

Coefficent:  [[-0.41571775  0.94192769 -2.46751316 -1.06071945]
 [ 0.55919397 -0.33376745 -0.22914878 -0.87704082]
 [-0.14347622 -0.60816025  2.69666194  1.93776027]]
Intercept:  [  9.64455738   2.13876161 -11.78331898]


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0          -0.415718          0.941928          -2.467513         -1.060719   
1           0.559194         -0.333767          -0.229149         -0.877041   
2          -0.143476         -0.608160           2.696662          1.937760   

   Intercept  
0   9.644557  
1   2.138762  
2 -11.783319

# Part 4: Support Vector Machine


In [358]:
# i. Use sklearn to train a Support Vector Classifier on the training set

svm_clf = svm.SVC(probability = True).fit(X_train,y_train.values.ravel())

# ii. For a sample datapoint, predict the probabilities for each possible class

## The same datapoint that I use above 
sample_datapoint = x[:1] 

##sample_datapoint with species 
sample_datapoint1 = df[:1]
sample_datapoint1

svm_prediction = svm_clf.predict_proba(sample_datapoint)[0]
print("Probability: ", svm_prediction[0], svm_prediction[1], svm_prediction[2])
data = {'Species': [0,1,2],
        'Species Name': ['Setosa', 'Versicolor', 'Virginica'],
        'Probability': ["{:.0%}".format(svm_prediction[0]), "{:.0%}".format(svm_prediction[1]), "{:.0%}".format(svm_prediction[2])]}
print(tabulate(data, headers="keys", tablefmt='psql'))

print("Total Probability: " , svm_prediction.sum())



Probability:  0.97392126003248 0.01660627695901868 0.009472463008501119
+-----------+----------------+---------------+
|   Species | Species Name   | Probability   |
|-----------+----------------+---------------|
|         0 | Setosa         | 97%           |
|         1 | Versicolor     | 2%            |
|         2 | Virginica      | 1%            |
+-----------+----------------+---------------+
Total Probability:  0.9999999999999999


In [359]:
# iii. Report on the score for the SVM, what does the score measure?

svm_train_score = svm_clf.score(X_train,y_train)
svm_test_score = svm_clf.score(X_test, y_test)
print("The train score for Support Vector model: ",svm_train_score)
print("The test score for Support Vector model ",svm_test_score)

The train score for Support Vector model:  0.9629629629629629
The test score for Support Vector model  1.0


 As we know, the score() function from sklearn will return the mean accuracy on the given test data and labels. The best possible you can get is 1, which means the model can pefectly predict the label. Therefore, if the score is really close to 1, it means your models is very good and accurate at predicting the label. 

Our Support Vector model got the score of 0.963 on the train date. The score is really close to 1, which means that our model is nearly perfect on predicting the species of the Iris during training. Based on the training data, the accuracy of this model is 96.3 % . 

Our Support Vector model got the score of 1 on the testing data. This means that our model perfectly predict the species of the Iris for all the testing dataset. This is impressive result that we get from our model. The score that we get indicate that our model is perfect on predicting the Iris species.

# Part 5: Neural Network


In [360]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set

nn_model = MLPClassifier(solver='lbfgs', alpha=1e-5, max_iter = 1000, activation = "tanh",
                     hidden_layer_sizes=(100,), random_state=1)
nn_model.fit(X_train, y_train.values.ravel())



MLPClassifier(activation='tanh', alpha=1e-05, max_iter=1000, random_state=1,
              solver='lbfgs')

In [361]:
# ii. For a sample datapoint, predict the probabilities for each possible class

## The same datapoint that I use above 
sample_datapoint = x[:1] 

##sample_datapoint with species 
sample_datapoint1 = df[:1]
sample_datapoint1

nnm_prediction = nn_model.predict_proba(sample_datapoint)[0]

print("Probability: ", nnm_prediction[0], nnm_prediction[1], nnm_prediction[2])
data = {'Species': [0,1,2],
        'Species Name': ['Setosa', 'Versicolor', 'Virginica'],
        'Probability': ["{:.0%}".format(nnm_prediction[0]), "{:.0%}".format(nnm_prediction[1]), "{:.0%}".format(nnm_prediction[2])]}
print(tabulate(data, headers="keys", tablefmt='psql'))

print("Total Probability: " , nnm_prediction.sum())


Probability:  1.0 2.0461584834053698e-26 1.2106788504769977e-86
+-----------+----------------+---------------+
|   Species | Species Name   | Probability   |
|-----------+----------------+---------------|
|         0 | Setosa         | 100%          |
|         1 | Versicolor     | 0%            |
|         2 | Virginica      | 0%            |
+-----------+----------------+---------------+
Total Probability:  1.0


In [362]:
# iii. Report on the score for the Neural Network, what does the score measure?
nnm_train_score = nn_model.score(X_train,y_train)
nnm_test_score = nn_model.score(X_test, y_test)
print("The train score for  Neural Network model: ",nnm_train_score)
print("The test score for  Neural Network model ",nnm_test_score)


The train score for  Neural Network model:  1.0
The test score for  Neural Network model  1.0


 As we know, the score() function from sklearn will return the mean accuracy on the given test data and labels. The best possible you can get is 1, which means the model can pefectly predict the label. Therefore, if the score is really close to 1, it means your models is very good and accurate at predicting the label. 

Our Neural Network model got the score of 1 on the train date. This means that our model perfectly predict the species of the Iris for all the training dataset. Based on the training data, the accuracy of this model is 100 % . 

Our Neural Network model got the score of 1 on the testing data. This means that our model perfectly predict the species of the Iris for all the testing dataset. This is impressive result that we get from our model. The score that we get indicate that our model is perfect on predicting the Iris species. Also, computer with logistic regression and SVM, this is the first time I get the perfect score on both of my dataset. This can prove neural network is very strong on predicting Iris species. 

### iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

The highest scores I got was 1 on both the training data and testing data 

-  The best configuation :
    - solver='lbfgs'
    - alpha=1e-5
    - max_iter = 1000
    - activation = "tanh"
    - hidden_layer_sizes=(100,)
    - random_state=1




# Part 6: K-Nearest Neighbors


In [363]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

neigh = KNeighborsClassifier(n_neighbors=1)
neigh.fit(X_train, y_train.values.ravel())


KNeighborsClassifier(n_neighbors=1)

In [364]:
# ii. For a sample datapoint, predict the probabilities for each possible class


## The same datapoint that I use above 
sample_datapoint = x[:1] 

##sample_datapoint with species 
sample_datapoint1 = df[:1]
sample_datapoint1

k_prediction = neigh.predict_proba(sample_datapoint)[0]

print("Probability: ", k_prediction[0], k_prediction[1], k_prediction[2])
data = {'Species': [0,1,2],
        'Species Name': ['Setosa', 'Versicolor', 'Virginica'],
        'Probability': ["{:.0%}".format(k_prediction[0]), "{:.0%}".format(k_prediction[1]), "{:.0%}".format(k_prediction[2])]}
print(tabulate(data, headers="keys", tablefmt='psql'))

print("Total Probability: " , k_prediction.sum())


Probability:  1.0 0.0 0.0
+-----------+----------------+---------------+
|   Species | Species Name   | Probability   |
|-----------+----------------+---------------|
|         0 | Setosa         | 100%          |
|         1 | Versicolor     | 0%            |
|         2 | Virginica      | 0%            |
+-----------+----------------+---------------+
Total Probability:  1.0


In [365]:
# iii. Report on the score for kNN, what does the score measure?

k_train_score = neigh.score(X_train,y_train)
k_test_score = neigh.score(X_test, y_test)
print("The train score for K-Nearest Neighbors model: ",k_train_score)
print("The test score for K-Nearest Neighbors model ",k_test_score)

The train score for K-Nearest Neighbors model:  1.0
The test score for K-Nearest Neighbors model  1.0


 As we know, the score() function from sklearn will return the mean accuracy on the given test data and labels. The best possible you can get is 1, which means the model can pefectly predict the label. Therefore, if the score is really close to 1, it means your models is very good and accurate at predicting the label. 

Our K-Nearest Neighbors model got the score of 0.96 on the train date. The score is really close to 1, which means that our model is nearly perfect on predicting the species of the Iris during training. Based on the training data, the accuracy of this model is 96.2 % . 

Our K-Nearest Neighbors model got the score of 1 on the testing data. This means that our model perfectly predict the species of the Iris for all the testing dataset. This is impressive result that we get from our model. The score that we get indicate that our model is perfect on predicting the Iris species.

# Part 7: Conclusions and takeaways


### In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

#### Model Summary 

| Model                 | Training Score     | Testing Score |
|-----------------------|--------------------|---------------|
| Logistic Regression   | 0.9777777777777777 | 1.0           |
| Suport Vector Machine | 0.9629629629629629 | 1.0           |
| Neural Network        | 1.0                | 1.0           |
| K-Nearest Neighbors   | 1.0                | 1.0           |


 * In this notebook, I use four different classification models on the Iris flower Dateset. I split the dataset into training data and testing data (90 % train and 10 test). The result is very surprised. After using the training data to train the model, the testing score for all four models are all 1.0, which means these four models are perfectly predict the Iris species based on their information about sepal length, sepal width , petal length and petal width. Even their trainijg score are all above 0.95 and two model have the prefect score on both training and testing. 

#### Which model performed the best on the dataset ?

* I think Neural Network and K-Nearest Neighbors perform the best on the dataset. The have the perfect scores on both the training data and testing score. Also, when I pick a datapoint to predict the result, they both have 100 % probability on choosing the correct species. 

#### Why do you think that is?

* I think neural network and k-nearest neighbor are non-linear classification models. If a problem is nonlinear and its class boundaries cannot be approximated well with linear hyperphanes, then nonlinear classifers are often more accurate than linear classifiers. In this case, I think the Iris flower dataset is nonlinear, so network and k-nearest will perform the best on the dataset. 


#### Did anything surprise you about the exercise?


 * One thing that surprise me is that all of the model are perfectly predict the testing data. This is so magical.
 * When I adjust the configuration of the k-nearest neighbor model, I find out that n_neighbors=1 will get the best result. However, when I increase the n_neighbors, it become worse and worse. 
 * I need to try so many time to find the right configuration in order for the model to have the best performance 